sync/sync.go

```diff
@@ -2,133 +2,176 @@
+		llog.Error("dial error in SendReport", llog.KV{"err": err})
 		return
 	}
-	startTransaction(req.Trans, serverAddr, now, req.Name)
-	send(Report, serverAddr, d)
+	t := InitTransaction(serverAddr, c)
+	//we need to start reading off the connection
+	go t.ReaderSpin()
+
+	go func() {
+		t.JobSpin()
+		c.Close()
+	}()
+	t.NewJob(&Job{
+		MType:   Init,
+		SrcAddr: serverAddr,
+	})
 }
 
-func HandleMessage(d []byte, srcAddr *net.UDPAddr) {
+func DecodeMessage(d []byte, srcAddr *net.UDPAddr) (*Job, error) {
 	if len(d) < 2 {
-		return
+		return nil, errors.New("Invalid message length")
 	}
 	var msg proto.Message
-	var fn func(proto.Message, *net.UDPAddr)
-	switch MsgType(d[0]) {
+	mt := MsgType(d[0])
+	switch mt {
 	case Report:
 		msg = &ReportRequest{}
-		fn = HandleReport
 	case Response:
 		msg = &ReportResponse{}
-		fn = HandleResponse
 	default:
-		llog.Warn("received invalid first byte", llog.KV{"byte": d[0], "src": srcAddr})
-		return
+		return nil, errors.New("received invalid first byte: " + strconv.Itoa(int(d[0])))
 	}
 	err := proto.Unmarshal(d[1:], msg)
 	if err != nil {
-		llog.Warn("unmarshaling error", llog.KV{"err": err, "src": srcAddr})
-		return
+		return nil, err
 	}
-	fn(msg, srcAddr)
+	return &Job{
+		SrcAddr: srcAddr,
+		MType:   mt,
+		Msg:     msg,
+	}, nil
 }
 
-func HandleReport(msg proto.Message, srcAddr *net.UDPAddr) {
-	req, ok := msg.(*ReportRequest)
-	if !ok {
-		return
+func getTransIDFromProto(mt MsgType, msg proto.Message) int32 {
+	switch mt {
+	case Response:
+		if r, ok := msg.(*ReportResponse); ok {
+			return r.Trans
+		}
 	}
-	llog.Debug("received req", llog.KV{"req": req, "addr": srcAddr})
-	if !verifyReport(req) {
-		llog.Warn("received unsigned report", llog.KV{"addr": srcAddr})
-		return
+	return 0
+}
+
+func EnsureTransaction(job *Job, conn *net.UDPConn) *Transaction {
+	var t *Transaction
+	switch job.MType {
+	case Report:
+		t = InitTransaction(job.SrcAddr, conn)
+		//start spinning for jobs
+		go t.JobSpin()
+	case Response:
+		t = getTransaction(job.SrcAddr.IP, getTransIDFromProto(job.MType, job.Msg))
 	}
+	return t
+}
 
-	now := time.Now()
-	if transactionExists(srcAddr.IP, req.Trans) {
-		llog.Warn("received duplicate report transactionID", llog.KV{"addr": srcAddr})
-		return
+func handleJob(j *Job, t *Transaction) []byte {
+	kv := llog.KV{"src": j.SrcAddr}
+	switch j.MType {
+	case Report:
+		if req, ok := j.Msg.(*ReportRequest); ok {
+			return handleReport(req, j.SrcAddr, t)
+		} else {
+			llog.Error("error casting to ReportRequest in handleJob", kv)
+		}
+	case Response:
+		if resp, ok := j.Msg.(*ReportResponse); ok {
+			return handleResponse(resp, j.SrcAddr, t)
+		} else {
+			llog.Error("error casting to ReportRequest in handleJob", kv)
+		}
 	}
-	if req.Seq != 1 {
-		llog.Warn("received report with 1 seq other than 1", llog.KV{"addr": srcAddr, "seq": req.Seq})
-		return
+	return nil
+}
+
+func handleReport(req *ReportRequest, srcAddr *net.UDPAddr, t *Transaction) []byte {
```

Comment:
For this and `handleResponse` you're not actually using `srcAddr` for anything except logging. Might be a bit cleaner to have them be able to return an error and let the caller do the error logging, instead of doing the logging themselves. The downside would be the debug logs would get less info though.....